Imgix Client Purge Method

tyrauber opened the pull request:
Implement basic token authentication api.imgix.com/v2/image/purger request using net:http and web mock for test suite

tyrauber committed:
Imgix Client Purge Method
Implement basic token authentication api.imgix.com/v2/image/purger request using net:http and web mock for test suite

tyrauber committed:
Update Gemfile to fix travis errors

tyrauber committed:
Update rbx version

jayeb commented on lib/imgix/client.rb:
```diff
@@ -24,6 +26,19 @@ def path(path)
     end
 
+    def purge(url)
+      raise "Authentication token required" unless !!(@secure_url_token)
```
I think if you test this you'll find it doesn't work--the `secure_url_token` you're using here is generated per-source, and is only used for signing imgix URLs generated from that source. The purge API method, on the other hand, requires the use of an imgix API token, which is generated per-user. We probably want to add the API token as a configuration value, alongside `host` and `secure_url_token`.

tyrauber replied:
Ahh, I see. I thought `@secure_url_token` was an odd attribute name for an api_token.  I added the `@api_token` initialization option and updated the purge method to utilize it. Nice catch.

jayeb commented on lib/imgix/client.rb:
```diff
@@ -24,6 +26,19 @@ def path(path)
+      uri = URI.parse('https://api.imgix.com/v2/image/purger')
+      req = Net::HTTP::Post.new(uri.path)
+      req.basic_auth @secure_url_token, ''
+      req.set_form_data({'url' => url})
```
The documentation you wrote made some mention of allowing the user to pass either a path or a fully-qualified URL, but I don't see the implementation following through on that. Personally I'd rather we only accept paths here, and always prepend the host. You can check out the `.to_url` method for some idea of how this should be implemented.

tyrauber replied:
Works for me. The implementation [here](https://github.com/imgix/imgix-rb/pull/37/commits/dcb41469bf1f15717e0995b14f3eb37b08cfa0e2#diff-b8db8534f93c103bbea4f27ab07961c1R31) was a simple string match on http, but was probably overkill.  Removed.

jayeb reviewed (changes requested):
Thanks for the PR, @tyrauber. Purge support is something we've been talking about adding to this library (and all of our core libraries) for a long time, but haven't had the time to do it just yet.

There's a couple of specific problems with the implementation at the moment, which I've noted in comments below. Happy to discuss any of these further if you have questions.

I also have a feature request that I'd like you to add if possible--we currently track how often our libraries are used via the [`ixlib` parameter](https://github.com/imgix/imgix-rb#what-is-the-ixlib-param-on-every-request), which we append to all URLs generated by each library unless the user turns the feature off. We'd like to do the same thing here, but with the `user-agent` header. Can you modify the code to add `User-Agent: imgix rb-VERSION` to each purge request?

tyrauber committed:
- add api_token initialization option
- use @api_token in purge request
- remove full url option
- add ixlib parameter to purge requests
- update purge_test.rb and README

tyrauber commented:
@jayeb Sorry for jumping the roadmap. Unfortunately, purge support is something we require in production ASAP. When I saw the curl purge example, I thought I might as well take the extra hour and fork and update the gem.

This last commit should resolve the outstanding concerns, but let me know.

FYI - I added the user agent as requested - `imgix @library-@version` - but it appears that [the rest of the gem only uses `@library-@version`](https://github.com/imgix/imgix-rb/blob/master/lib/imgix/client.rb#L23).  Can you confirm imgix should also be included in the user agent? If so, the core path method should probably also be updated.

jayeb commented on lib/imgix/client.rb:
```diff
@@ -11,6 +13,7 @@ def initialize(options = {})
 
       @hosts = Array(options[:host]) + Array(options[:hosts]) and validate_hosts!
       @secure_url_token = options[:secure_url_token]
+      @api_token = options[:api_token]
```
Sorry, I just checked and in our docs we use the term "API Key", not "API token". My bad. For consistency's sake, can we update this variable and option name?

jayeb commented on lib/imgix/client.rb:
```diff
@@ -24,6 +27,20 @@ def path(path)
+      url = prefix(path)+path
+      uri = URI.parse('https://api.imgix.com/v2/image/purger')
+      req = Net::HTTP::Post.new(uri.path)
+      req.initialize_http_header({"User-Agent" => "imgix #{@library}-#{@version}"})  if @include_library_param
```
Let's ditch the `if @include_library_param` here. We allow users to turn that off because it's sometimes important for our users to have complete control over what they ship to *their* users, which means letting them turn off the "extraneous" `ixlib` parameter when generating URLs. In the case of purging, though, the only parties involved in a purge request are our direct user (whoever is implementing this library in their code) and imgix, so the ability to turn off this tracking header doesn't benefit anyone.

---

In answer to your other question, we *do* want the "imgix" prefix here but we don't want it for the `ixlib` param. In both cases, we're adding some metadata to the request to allow us to easily answer two questions about any given render or purge request:

1. Was this request generated by an imgix library?
2. If so, which library generated it?

For render requests, we can answer question 1 by looking for the presence of the `ixlib` parameter, and we answer question 2 by looking at that parameter's value. For purge requests, we'll be able to look for the `imgix` prefix in the `User-Agent` header to answer question 1 and everything after the prefix to answer question 2.

jayeb commented on test/units/purge_test.rb:
```diff
@@ -0,0 +1,25 @@
+    
+    assert_requested :post, 'https://api.imgix.com/v2/image/purger',
+      body:  'url=https%3A%2F%2Fdemo.imgix.net%2Fimages%2Fdemo.png',
+      headers: {'Accept'=>'*/*', 'Accept-Encoding'=>'gzip;q=1.0,deflate;q=0.6,identity;q=0.3', 'Authorization'=>'Basic MTBhZGMzOTQ6', 'Content-Type'=>'application/x-www-form-urlencoded', 'User-Agent'=>'Ruby'},
```
Any reason this test can't check that the correct `User-Agent` header is provided as well?

jayeb commented on README.md:
```diff
@@ -78,6 +78,19 @@ For example to use the noise reduction:
+
+## Purge Cache
+
+If you need to remove or update an image on imgix, you can purge it from the cache with the following method.
```
Linking to our docs here would be helpful, I think. How about this?

> If you need to remove or update an image on imgix, you can purge it from our cache by initializing a client with your `api_key`, then calling `Imgix::Client#purge` with the resource path.
> ``` ruby
> client = Imgix::Client.new(host: 'your-subdomain.imgix.net', api_key: 'your-key')
> client.purge('/images/demo.png')
> ```
> To learn more about purging assets with imgix, [see our docs](https://docs.imgix.com/setup/purging-images).

jayeb reviewed (changes requested):
A few more fixes, @tyrauber. Thanks for sticking with it.

tyrauber committed:
- api_token changed to api_key
- remove ability to disable library tracking
- test should test user-agent
- update README content

tyrauber commented:
Of course, @jayeb, Glad to help.  Let me know if you need anything else.

benhutton commented on imgix.gemspec:
```diff
@@ -20,4 +20,6 @@ Gem::Specification.new do |spec|
 
   spec.required_ruby_version = '>= 1.9.0'
   spec.add_dependency 'addressable'
+  spec.add_dependency 'http'
```
sorry if I'm being dense, but is this gem actually being used here? AFAIK, it uses an `HTTP` class, and this PR only uses `Net::HTTP`, not `HTTP`. It'd be great to not require it if we don't really need it!

tyrauber replied:
@benhutton You are absolutely correct. Addressed in #38. Thanks for catching that.
